mix.exs

```diff
@@ -34,6 +34,7 @@ defmodule Plug.Cowboy.MixProject do
   def deps do
     [
       {:plug, "~> 1.7"},
+      {:sock, "~> 0.3.0"},
```

Comment:
What do you think of making this part of plug as `Plug.Websocket` instead? Not a fan of adding a dependency that is just a single behaviour when we already have plug for HTTP interfaces.

Reply:
Yes, I dropped a comment. We should not depend on Sock because we should not expose the Sock API but the Cowboy API. Then Sock sits on top!

Reply:
~Got it.~ Actually, maybe I don't :)

I follow you as far as removing Sock from this PR, and having Plug.Cowboy expose Cowboy WS handler API. That much is clear.

When you say 'Sock sits on top!', are you suggesting that a caller does something like:

def call(conn, opts) do
   ....
   # Assuming that MyApp.Socket is what you ultimately want to handle the WS
   upgrade_adapter(conn, :websocket, {Sock, {MyApp.Socket, opts}, connection_opts})
end
At which time the underlying server (say Cowboy) will call Sock via the Cowboy WS API, and Sock will then call onwards to MyApp.Socket via the Sock API?

Reply:
Yes, the goal is to have `Sock.upgrade(conn, sock_options)` and that will call `upgrade_adapter` for each adapter accordingly.

Reply:
Got it; Sock in this case would wrap the socket lifecycle starting from upgrade. 

Only thing I'm not sure is how that would look in the first PR split from phoenixframework/phoenix#5030 before we introduce Sock. Pretty sure it would still need to be cowboy specific.

Reply:
Yeah, let’s not worry about Sock in that PR yet. Let’s change the plug dispatch bits and I think figuring out the rest will be straightforward!